cfgmgr/buffermgrdyn.cpp

```diff
@@ -0,0 +1,1620 @@
+
+            //Try updating the buffer information
+            switch (portInfo.state)
+            {
+            case PORT_INITIALIZING:
+                portInfo.state = PORT_READY;
+                if (mtu.empty())
+                {
+                    // During initialization, the flow can be
+                    // 1. mtu, cable_length, speed
+                    // 2. cable_length, speed, mtu
+                    // 3. cable_length, speed, but no mtu specified
+                    // it's ok for the 1st case
+                    // but for the 2nd case, we can't wait mtu for calculating the headrooms
+                    // because we can't distinguish case 2 from case 3 which is also legal.
+                    // So once we have speed updated, let's try to calculate the headroom with default mtu.
+                    // The cost is that if the mtu is provided in the following iteration
+                    // the current calculation is of no value and will be replaced.
+                    mtu = DEFAULT_MTU_STR;
+                }
+                task_status = refreshPriorityGroupsForPort(port, speed, cable_length, mtu);
+                break;
+
+            case PORT_READY:
+                task_status = refreshPriorityGroupsForPort(port, speed, cable_length, mtu);
+                break;
+            }
+
+            SWSS_LOG_DEBUG("Port Info for %s after handling speed %s cable %s gb %s",
+                           port.c_str(),
+                           portInfo.speed.c_str(), portInfo.cable_length.c_str(), portInfo.gearbox_model.c_str());
+        }
+        else if (admin_status_updated)
+        {
+            SWSS_LOG_INFO("Recalculate shared buffer pool size due to port %s's admin_status updated", port.c_str());
+            checkSharedBufferPoolSize();
+        }
+    }
+
+    return task_status;
+}
+
+task_process_status BufferMgrDynamic::handleBufferPoolTable(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+    auto it = consumer.m_toSync.begin();
+    KeyOpFieldsValuesTuple tuple = it->second;
+    string &pool = kfvKey(tuple);
+    string op = kfvOp(tuple);
+    vector<FieldValueTuple> fvVector;
+
+    SWSS_LOG_DEBUG("processing command:%s table BUFFER_POOL key %s", op.c_str(), pool.c_str());
+    if (op == SET_COMMAND)
+    {
+        // For set command:
+        // 1. Create the corresponding table entries in APPL_DB
+        // 2. Record the table in the internal cache m_bufferPoolLookup
+        buffer_pool_t &bufferPool = m_bufferPoolLookup[pool];
+
+        bufferPool.initialized = false;
+        bufferPool.dynamic_size = true;
+        for (auto i = kfvFieldsValues(tuple).begin(); i != kfvFieldsValues(tuple).end(); i++)
+        {
+            string &field = fvField(*i);
+            string &value = fvValue(*i);
+
+            SWSS_LOG_DEBUG("field:%s, value:%s", field.c_str(), value.c_str());
+            if (field == buffer_size_field_name)
+            {
+                bufferPool.dynamic_size = false;
```

Comment:
Should we check the buffer pool size for any error if this is more than max mmu size??

Reply:
Done.